src/Microsoft.OData.Core/Json/JsonReaderExtensions.cs

```diff
@@ -459,5 +905,45 @@ private static void ValidateNodeType(this IJsonReader jsonReader, JsonNodeType e
                 throw CreateException(Strings.JsonReaderExtensions_UnexpectedNodeDetected(expectedNodeType, jsonReader.NodeType));
             }
         }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the <paramref name="jsonReader"/> and verifies that it is of the expected node type.
+        /// </summary>
+        /// <param name="jsonReader">The <see cref="JsonReader"/> to read from.</param>
+        /// <param name="expectedNodeType">The expected <see cref="JsonNodeType"/> of the read node.</param>
+        private static async Task ReadNextAsync(this IJsonReaderAsync jsonReader, JsonNodeType expectedNodeType)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+            Debug.Assert(expectedNodeType != JsonNodeType.None, "expectedNodeType != JsonNodeType.None");
+
+            jsonReader.ValidateNodeType(expectedNodeType);
+            await jsonReader.ReadAsync().ConfigureAwait(false);

```

Comment:
```suggestion
        private static Task ReadNextAsync(this IJsonReaderAsync jsonReader, JsonNodeType expectedNodeType)
        {
            Debug.Assert(jsonReader != null, "jsonReader != null");
            Debug.Assert(expectedNodeType != JsonNodeType.None, "expectedNodeType != JsonNodeType.None");

            jsonReader.ValidateNodeType(expectedNodeType);
            return jsonReader.ReadAsync().ConfigureAwait(false);

```

Reply:
@odero Eliding may not be a good idea in this method since `ValidateNodeType` does throw an exception when the node type is not the expected one. The state machine will allow us to stash the exception in the returned task